### PR TITLE
[EC-90] Persistent node keys

### PR DESCRIPTION
**Pull request by ntallar**
## Description

Includes use of `nodeId.keys` file to persist the public and private keys (as plain text for now).

### CODE REVIEW

File: src/main/resources/application.conf [hunk 1]
```diff
@@ -2,6 +2,8 @@ etc-client {
 
   client-id = "etc-client"
 
+  keys-file = "nodeId.keys"
```
Comment (whysoserious) [inline src/main/resources/application.conf]: Could you add some docs explaining syntax of a `keys-file`? This is not a blocker.

File: src/main/scala/io/iohk/ethereum/network/package.scala [hunk 2]
```diff
@@ -19,4 +25,38 @@ package object network {
       Hex.decode(nodeId)
     curve.getCurve.decodePoint(bytes)
   }
+
+  object AsymmetricCipherKeyPairSerializable {
+
+    def toBytes(keysValuePair: AsymmetricCipherKeyPair): (String, String) = {
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/network/package.scala]: `toBytes` -> `toHexStrings` ?

File: src/main/scala/io/iohk/ethereum/network/package.scala [hunk 3]
```diff
@@ -19,4 +25,38 @@ package object network {
...
+      val pubKey = keysValuePair.getPublic.asInstanceOf[ECPublicKeyParameters].toNodeId
+      val privKey = keysValuePair.getPrivate.asInstanceOf[ECPrivateKeyParameters].getD.toByteArray
+      (Hex.toHexString(pubKey), Hex.toHexString(privKey))
+    }
+
+    def fromBytes(publicKey: String, privateKey: String): AsymmetricCipherKeyPair = {
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/network/package.scala]: `fromBytes` -> `fromHexStrings` ?

File: src/main/scala/io/iohk/ethereum/network/package.scala [hunk 4]
```diff
@@ -19,4 +25,38 @@ package object network {
       Hex.decode(nodeId)
     curve.getCurve.decodePoint(bytes)
   }
+
```
Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/network/package.scala]: This probably deserves it's own object in ti's own file.And I think it's orthogonal to network. Possibly 'crypto' or 'utils'

**Comment by rtkaczyk** [comment]
Looks good, but I would change those names. I see the word `bytes`, but the type is `String` and I get confused 😉

File: src/main/scala/io/iohk/ethereum/network/package.scala [hunk 5]
```diff
@@ -19,4 +25,38 @@ package object network {
...
+  }
+
+  def loadAsymmetricCipherKeyPair(filePath: String): AsymmetricCipherKeyPair = {
+    if(!new File(filePath).exists()){
+      val keysValuePair = generateKeyPair()
+
+      //Write keys to file
+      val (pub, priv) = AsymmetricCipherKeyPairSerializable.toBytes(keysValuePair)
+      val file = new PrintWriter(filePath)
+      file.write(pub ++ "\n" ++ priv)
```
Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/network/package.scala]: Can we load this as a system property? (I believe so)

File: src/test/scala/io/iohk/ethereum/network/AsymmetricCipherKeyPairSerializableSpec.scala [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    privateKeyParam.getD shouldBe privateKeyParamObtained.getD
+    privateKeyParam.getParameters shouldBe privateKeyParamObtained.getParameters
+    privateKeyParam.isPrivate shouldBe privateKeyParamObtained.isPrivate
+
+    //Serializing
+    val (pub2, priv2) = AsymmetricCipherKeyPairSerializable.toBytes(keysValuePairObtained)
+    ByteString(pub2) shouldBe ByteString(pub)
+    ByteString(priv2) shouldBe ByteString(priv)
+  }
+}
```
Comment (mcsherrylabs) [inline src/test/scala/io/iohk/ethereum/network/AsymmetricCipherKeyPairSerializableSpec.scala]: :+1:

File: src/main/scala/io/iohk/ethereum/network/package.scala [hunk 7]
```diff
@@ -19,4 +25,38 @@ package object network {
...
+      file.close()
+
```
Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/network/package.scala]: Should file.close happen in a 'finally'?

File: src/main/scala/io/iohk/ethereum/network/package.scala (hunk 5 above)
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/network/package.scala]: my concern is that we are storing private key as plain text, but I guess there is no other option

Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/network/package.scala]: We *could* encrypt it with a passphrase. But not today.... :)

Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/network/package.scala]: @adamsmo me too but I guess this task was only for storing this in plain text. That being said +1 to address this in the future as @mcsherrylabs said :smile:

File: build.sbt [hunk 8]
```diff
@@ -38,4 +38,7 @@ val root = project.in(file("."))
 scalacOptions := Seq("-unchecked", "-deprecation", "-feature")
 
 (scalastyleConfig in Test) := baseDirectory.value / "scalastyle-test-config.xml"
-scalastyleSources in Test ++= {(unmanagedSourceDirectories in Integration).value}
\ No newline at end of file
+scalastyleSources in Test ++= {(unmanagedSourceDirectories in Integration).value}
+
+fork := true
+javaOptions += "-Dkey-file-format=%s\n%s"
```
Comment (adamsmo) [inline build.sbt]: minor: new line at end of the file

File: src/main/scala/io/iohk/ethereum/network/package.scala [hunk 9]
```diff
@@ -19,4 +25,26 @@ package object network {
       Hex.decode(nodeId)
     curve.getCurve.decodePoint(bytes)
   }
+
+  def loadAsymmetricCipherKeyPair(filePath: String): AsymmetricCipherKeyPair = {
```
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/network/package.scala]: maybe I am missing something but this method name is `loadAsymmetricCipherKeyPair` but this method writes keys to file `file.write(keyFileFormat.format(pub, priv))` should we change the name?

Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/network/package.scala]: ignore this comment

File: src/main/scala/io/iohk/ethereum/network/package.scala [hunk 10]
```diff
@@ -19,4 +25,26 @@ package object network {
...
+      try {
+        val keyFileFormat: String = System.getProperty("key-file-format")
+        file.write(keyFileFormat.format(pub, priv))
+      } finally {
+        file.close()
+      }
+
+      keysValuePair
+    } else {
+      val List(pub, priv) = Source.fromFile(filePath).getLines().toList
```
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/network/package.scala]: what will happen if I change file format from `-Dkey-file-format=%s\n%s` for example to -Dkey-file-format=%s %s

File: build.sbt (hunk 8 above)
Comment (whysoserious) [inline build.sbt]: We keep all our configuration in `application.conf` Why did we decide to read this particular setting from `javaOptions`?

File: src/main/scala/io/iohk/ethereum/network/package.scala (hunk 10 above)
Comment (whysoserious) [inline src/main/scala/io/iohk/ethereum/network/package.scala]: ^^

File: build.sbt (hunk 8 above)
Comment (whysoserious) [inline build.sbt]: Also, what's the usecase for making this format configurable?

**Comment by ntallar** [comment]
@whysoserious Having the key format configurable was required by @mcsherrylabs in:

> Can we load this as a system property? (I believe so)

File: src/main/scala/io/iohk/ethereum/network/package.scala (hunk 10 above)
Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/network/package.scala]: As the guys have said I don't think we need a configurable format for this file. If I gave you a bad steer on this at some stage - apologies.

Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/network/package.scala]: No problem, I can change it back quickly :+1:

Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/network/package.scala]: @ntallar I meant the CR/LF character could be loaded from system properties! I'll get an example...

**Comment by whysoserious** [comment]
LGTM

**Pull request merged**